src/ray/protobuf/gcs.proto

```diff
@@ -453,3 +356,103 @@ message PubSubMessage {
+// backward compatibility in autoscaler.
+
+// Represents the demand for a particular resource shape.
+message ResourceDemand { // Do not change!
+  // The resource shape requested. This is a map from the resource string
+  // (e.g., "CPU") to the amount requested.
+  map<string, double> shape = 1; // Do not change!
+  // The number of requests that are ready to run (i.e., dependencies have been
+  // fulfilled), but that are waiting for resources.
+  uint64 num_ready_requests_queued = 2; // Do not change!
+  // The number of requests for which there is no node that is a superset of
+  // the requested resource shape.
+  uint64 num_infeasible_requests_queued = 3; // Do not change!
+  // The number of requests of this shape still queued in CoreWorkers that this
+  // raylet knows about.
+  int64 backlog_size = 4; // Do not change!
+} // Do not change!
+
+// Represents the demand sorted by resource shape.
+message ResourceLoad { // Do not change!
+  // A list of all resource demands. The resource shape in each demand is
+  // unique.
+  repeated ResourceDemand resource_demands = 1; // Do not change!
+} // Do not change!
+
+message ResourcesData { // Do not change!
+  // Node id.
+  bytes node_id = 1; // Do not change!
+  // Resource capacity currently available on this node manager.
+  map<string, double> resources_available = 2; // Do not change!
+  // Indicates whether available resources is changed. Only used when light
+  // heartbeat enabled.
+  bool resources_available_changed = 3; // Do not change!
+  // Total resource capacity configured for this node manager.
+  map<string, double> resources_total = 4; // Do not change!
+  // Aggregate outstanding resource load on this node manager.
+  map<string, double> resource_load = 5; // Do not change!
+  // Indicates whether resource load is changed. Only used when
+  // light heartbeat enabled.
+  bool resource_load_changed = 6; // Do not change!
+  // The resource load on this node, sorted by resource shape.
+  ResourceLoad resource_load_by_shape = 7; // Do not change!
+  // Whether this node manager is requesting global GC.
+  bool should_global_gc = 8; // Do not change!
+  // IP address of the node.
+  string node_manager_address = 9; // Do not change!
+} // Do not change!
+
+message ResourceUsageBatchData { // Do not change!
+  repeated ResourcesData batch = 1; // Do not change!
+  // The total resource demand on all nodes included in the batch, sorted by
+  // resource shape.
+  ResourceLoad resource_load_by_shape = 2; // Do not change!
+  // The pending list of placement groups.
+  PlacementGroupLoad placement_group_load = 3; // Do not change!
+} // Do not change!
+
+message PlacementGroupLoad { // Do not change!
+  // The list of pending placement group specifications.
+  repeated PlacementGroupTableData placement_group_data = 1; // Do not change!
+} // Do not change!
+
+message PlacementGroupTableData { // Do not change!
```

Comment:
This is also hard to say it shouldn't be changed because it is literally the implementation detail of placement groups. Why don't we just parse necessary fields from these instead of marking them do not change?

Reply:
If you change any field name here, or the. Name of the table, or the type of it, or add something crucial for autoscaling, it will break autoscaler/monitor.

Reply:
I agree with @rkooo567 here, it's not professional to add this everywhere. Let's add it one at the top of the message with a reason why.

Reply:
I moved everything to autoscaler.proto